com.zsmartsystems.zigbee/src/main/java/com/zsmartsystems/zigbee/ZigBeeNetworkManager.java

```diff
@@ -1389,24 +1399,29 @@ public void updateNode(final ZigBeeNode node) {
             nodeDiscoveryComplete.add(node.getIeeeAddress());
         }
 
-        synchronized (this) {
-            for (final ZigBeeNetworkNodeListener listener : nodeListeners) {
-                NotificationService.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        if (updated) {
-                            listener.nodeUpdated(currentNode);
-                        } else {
-                            listener.nodeAdded(currentNode);
-                        }
+        for (final ZigBeeNetworkNodeListener listener : nodeListeners) {
+            NotificationService.execute(new Runnable() {
+                @Override
+                public void run() {
+                    if (updated) {
+                        listener.nodeUpdated(currentNode);
+                    } else {
+                        listener.nodeAdded(currentNode);
                     }
-                });
-            }
+                }
+            });
+        }
 
-            if (networkStateSerializer != null) {
-                networkStateSerializer.serialize(this);
+        executorService.execute(new Runnable() {
+            @Override
+            public void run() {
+                synchronized (ZigBeeNetworkManager.this) {
+                    if (networkStateSerializer != null) {
```

Comment:
Wouldn't it be better to put this check outside of the thread?

Reply:
Same as above...